File: src/Core/CommandBase.cs
using System;
using System.Collections.Generic;
using Mono.Options;

namespace NPackage.Core
{
    public abstract class CommandBase : ICommand
    {
        private readonly List<string> extraArguments = new List<string>();

        protected virtual OptionSet CreateOptionSet()
        {
            return new OptionSet().Add("h|help", "show this message and exit", v => ShowHelp = v != null);
        }

        public void ParseOptions(IEnumerable<string> arguments)
        {
            OptionSet set = CreateOptionSet();
            extraArguments.AddRange(set.Parse(arguments));
        }

        protected abstract int RunCore();

        public int Run()
        {
            if (ShowHelp)
            {
                Console.WriteLine("Usage:");
                CreateOptionSet().WriteOptionDescriptions(Console.Out);
                return 0;
            }
            else
                return RunCore();
        }

        protected IList<string> ExtraArguments
        {
            get { return extraArguments; }
        }

        public bool ShowHelp { get; set; }
    }
}

File: src/Core.FSharp/CommandBase.fs
namespace NPackage.Core

open System
open System.Collections.Generic
open Mono.Options

[<AbstractClass>]
type CommandBase() =
    let extraArguments = new ResizeArray<string>()
    let mutable showHelp = false

    abstract CreateOptionSet : unit -> OptionSet
    abstract RunCore : unit -> int

    default this.CreateOptionSet() =
        (new OptionSet())
            .Add("h|help", "show this message and exit", function
                                                         | null -> showHelp <- false
                                                         | _ -> showHelp <- true)


    member this.ExtraArguments = extraArguments :> IList<string>

    member this.ShowHelp
        with get() = showHelp
        and set(value) = showHelp <- value

    interface ICommand with
        member this.ParseOptions(arguments) =
            let set = this.CreateOptionSet()
            extraArguments.AddRange(set.Parse(arguments))

        member this.Run() =
            if showHelp then
                printfn "Usage:"
                this.CreateOptionSet().WriteOptionDescriptions(Console.Out)
                0
            else
                this.RunCore()
